### PR TITLE
Feature: Add basic Google Artifact Registry support

### PR DESCRIPTION
**Pull request by patricklubach**
# Pull Request

## Description

This PR adds basic support for Artifact Registry as on the one hand I need the support and on the other GCR Cleaner supports Artifact Registry too.

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## How Has This Been Tested?

I created a new Google Cloud project, switched to examples/minimal, created a terraform.tfvars file, add all necessary values for required variables. Run `terraform init`, `terraform plan -out=terraform.plan` and `terraform apply terraform.plan`.
When deployment finishes I verify GCR Cleaner Cloud Run and Cloud Scheduler Trigger are there. Trigger Cloud Run triggers and check status of last run was successful.

## Checklist:

- [x] My PR title follows [semantics prefix](.github/CONTRIBUTING.md#semantic-pull-requests)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked my code and corrected any misspellings
- [x] I have run pre-commit hooks `pre-commit run -a`
- [x] CI tests are passing

### CODE REVIEW

File: iam.tf [hunk 1]
```diff
@@ -36,3 +36,19 @@ resource "google_project_iam_member" "this" {
...
+# and delete container images.
+resource "google_artifact_registry_repository_iam_member" "this" {
+  for_each = {
+    for repo in var.gar_repositories : repo.name => repo
+  }
+
+  project    = each.value.project_id != null ? each.value.project_id : local.google_project_id
+  location   = each.value.region
+  repository = "projects/${each.value.project_id != null ? each.value.project_id : local.google_project_id}/locations/${each.value.region}/repositories/${each.value.name}"
+  role       = "roles/artifactregistry.writer"
```
Comment (anouarchattouna) [inline iam.tf:50]:
```suggestion
  role       = "roles/artifactregistry.repoAdmin"
```

Comment (anouarchattouna) [reply iam.tf:50]: It seems that the `writer` role is missing the `artifactregistry.repositories.deleteArtifacts` permission !

Comment (anouarchattouna) [reply iam.tf:50]: see https://cloud.google.com/artifact-registry/docs/access-control#roles

File: variables.tf [hunk 2]
```diff
@@ -121,6 +121,40 @@ EOF
...
+        keep           = If an integer is provided, it will always keep that minimum number of images. Note that it will not consider images inside the `grace` duration (optional(string))
+        tag_filter     = (Deprecated) If specified, any image where the first tag matches this given regular expression will be deleted. The image will not be deleted if other tags match the regular expression (optional(string))
+        tag_filter_any = If specified, any image with at least one tag that matches this given regular expression will be deleted. The image will be deleted even if it has other tags that do not match the given regular expression (optional(string))
+        tag_filter_all = If specified, any image where all tags match this given regular expression will be deleted. The image will not be delete if it has other tags that do not match the given regular expression (optional(string))
+    })))
```
Comment (anouarchattouna) [inline variables.tf:138]:
```suggestion
        dry_run        = If set to true, will not delete anything and outputs what would have been deleted. (optional(bool))
    })))
```

File: variables.tf [hunk 3]
```diff
@@ -121,6 +121,40 @@ EOF
...
+  type = list(object({
+    project_id = optional(string)
+    region     = string
+    name       = string
+    parameters = optional(object({
+      grace          = optional(string)
+      keep           = optional(string)
+      tag_filter     = optional(string)
+      tag_filter_any = optional(string)
+      tag_filter_all = optional(string)
```
Comment (anouarchattouna) [inline variables.tf:152]:
```suggestion
      tag_filter_all = optional(string)
      dry_run        = optional(bool)
```

**Review by anouarchattouna — CHANGES_REQUESTED** [review]
Could you please add the `google-beta` provider configuration to each example's `versions.tf` file ?
```hcl
    google-beta = {
      source  = "hashicorp/google-beta"
      version = ">= 4.1.0"
    }
   ```

**Comment by patricklubach** [comment]
> Could you please add the `google-beta` provider configuration to each example's `versions.tf` file ?
> 
> ```terraform
>     google-beta = {
>       source  = "hashicorp/google-beta"
>       version = ">= 4.1.0"
>     }
> ```

I added them. Please review :)

**Review by anouarchattouna — APPROVED** [review]
LGTM!
Could you just double-check the `pre-commit run -a` command

File: locals.tf [hunk 4]
```diff
@@ -66,6 +66,21 @@ locals {
...
+      keep           = gar.parameters != null ? (gar.parameters.keep != null ? gar.parameters.keep : "0") : "0"
+      tag_filter     = gar.parameters != null ? (gar.parameters.tag_filter != null ? gar.parameters.tag_filter : "") : ""
+      tag_filter_any = gar.parameters != null ? (gar.parameters.tag_filter_any != null ? gar.parameters.tag_filter_any : "") : ""
+      tag_filter_all = gar.parameters != null ? (gar.parameters.tag_filter_all != null ? gar.parameters.tag_filter_all : "") : ""
+      recursive      = false
+      dry_run        = gar.parameters != null ? (gar.parameters.dry_run != null ? gar.parameters.dry_run : false) : false
+      filter         = gar.parameters != null ? "grace-${gar.parameters.grace != null ? gar.parameters.grace : "0"}-keep-${gar.parameters.keep != null ? gar.parameters.keep : "0"}-tag_filter-${gar.parameters.tag_filter != null ? gar.parameters.tag_filter : "no"}-tag_filter_any-${gar.parameters.tag_filter_any != null ? gar.parameters.tag_filter_any : "no"}-tag_filter_any-${gar.parameters.tag_filter_any != null ? gar.parameters.tag_filter_any : "no"}" : "delete-all-untagged-images-recursive"
```
Comment (PascalBourdier) [inline locals.tf:80]: typo ?
```suggestion
      filter         = gar.parameters != null ? "grace-${gar.parameters.grace != null ? gar.parameters.grace : "0"}-keep-${gar.parameters.keep != null ? gar.parameters.keep : "0"}-tag_filter-${gar.parameters.tag_filter != null ? gar.parameters.tag_filter : "no"}-tag_filter_any-${gar.parameters.tag_filter_any != null ? gar.parameters.tag_filter_any : "no"}" : "delete-all-untagged-images-recursive"
```

Comment (patricklubach) [reply locals.tf]: Yes it was a typo but I missed `tag_filter_all`. Thanks for findings it! I fixed it.

**Review by PascalBourdier — APPROVED** [review]

**Pull request merged**